wfe2/verify.go

```diff
@@ -5,246 +5,540 @@ import (
-		key = submittedKey
-	} else if err != nil {
-		// For all other errors, or if regCheck is true, return error immediately.
-		wfe.stats.Inc("Errors.UnableToGetRegistrationByKey", 1)
-		logEvent.AddError("unable to fetch registration by the given JWK: %s", err)
+		wfe.stats.Inc("Errors.InvalidKeyID", 1)
+		logEvent.AddError("JWS key ID was invalid int64")
+		return nil, nil, probs.Malformed(fmt.Sprintf("Malformed account ID in KeyID header"))
+	}
+
+	// Try to find the account for this account ID
+	account, err := wfe.SA.GetRegistration(ctx, accountID)
+	if err != nil {
+		// If the account isn't found, return a suitable problem
 		if berrors.Is(err, berrors.NotFound) {
-			return nil, nil, reg, probs.Unauthorized(unknownKey)
+			wfe.stats.Inc("Errors.KeyIDNotFound", 1)
+			logEvent.AddError(fmt.Sprintf("Account %q not found", accountURL))
+			return nil, nil, probs.AccountDoesNotExist(fmt.Sprintf(
+				"Account %q not found", accountURL))
 		}
 
-		return nil, nil, reg, probs.ServerInternal("Failed to get registration by key")
-	} else {
-		// If the lookup was successful, use that key.
-		key = reg.Key
-		logEvent.Requester = reg.ID
-		logEvent.Contacts = reg.Contact
+		// If there was an error and it isn't a "Not Found" error, return
+		// a ServerInternal problem since this is unexpected.
+		wfe.stats.Inc("Errors.UnableToGetAccountByID", 1)
+		logEvent.AddError(fmt.Sprintf("Error calling SA.GetRegistration: %s", err.Error()))
+		return nil, nil, probs.ServerInternal(fmt.Sprintf(
+			"Error retreiving account %q", accountURL))
 	}
 
-	// Only check for validity if we are actually checking the registration
-	if regCheck && features.Enabled(features.AllowAccountDeactivation) && reg.Status != core.StatusValid {
-		return nil, nil, reg, probs.Unauthorized(fmt.Sprintf("Registration is not valid, has status '%s'", reg.Status))
+	// Verify the account is not deactivated
+	if account.Status != core.StatusValid {
+		wfe.stats.Inc("Errors.AccountIsNotValid", 1)
+		logEvent.AddError(fmt.Sprintf("Account %q has status %q", accountURL, account.Status))
+		return nil, nil, probs.Unauthorized(
+			fmt.Sprintf("Account is not valid, has status %q", account.Status))
 	}
 
-	if statName, err := checkAlgorithm(key, parsedJws); err != nil {
+	// Update the logEvent with the account information and return the JWK
+	logEvent.Requester = account.ID
+	logEvent.Contacts = account.Contact
+	return account.Key, &account, nil
+}
+
+// validJWSForKey checks a provided JWS for a given HTTP request validates
+// correctly using the provided JWK. If the JWS verifies the protected payload
+// is returned. The key/JWS algorithms are verified and
+// the JWK is checked against the keyPolicy before any signature validation is
+// done. If the JWS signature validates correctly then the JWS nonce value
+// and the JWS URL are verified to ensure that they are correct.
+func (wfe *WebFrontEndImpl) validJWSForKey(
+	jws *jose.JSONWebSignature,
+	jwk *jose.JSONWebKey,
+	body string,
```

Comment:
It looks like `parseJWS` returns a body (payload), and `validJWSForKey` takes it as an argument, for the sole reason of being able to print part of the payload in the case of a validation error. I think it would be cleaner to remove that from the error message, and not touch the payload until after validation. Or to put it another way, the only way to get a copy of the payload should be from the return value of `validJWSForKey`.